### PR TITLE
Flatten TRITONCACHE API structure, use callbacks to avoid unnecessary copies

### PR DESCRIPTION
**Pull request by rmccorm4**
Corresponding local_cache pr: https://github.com/triton-inference-server/local_cache/pull/4

### CODE REVIEW

File: src/cache_entry.cc [hunk 1]
```diff
@@ -32,123 +32,305 @@ namespace triton { namespace core {
...
+
+  std::byte* base = static_cast<std::byte*>(buffer.first);
+  if (!base) {
+    return Status(Status::Code::INTERNAL, "buffer was nullptr");
+  }
+
+  uint64_t position = 0;
+  // Number of response outputs serialized in buffer
+  uint32_t num_outputs = 0;
+  std::memcpy(&num_outputs, base, sizeof(num_outputs));
```
[the remark this answers is not in the document]
Comment (rmccorm4) [reply src/cache_entry.cc:323]: @Tabrizian Can you clarify which lines are copying twice? 

~Line 169: `SerializeResponse` (called through ResponseToCacheAllocator callback) will copy from response outputs into packed buffer.

~Line 318: `DeserializeBuffer` (called through CacheToResponseAllocator callback) will copy from packed buffer into local vars for parsing and response outputs

I believe the only redundancy/duplication is that `GetByteSize` will read the values from each `InferenceResponse::Output::DataBuffer` to get the metadata (output size, dtype, shape, etc) and compute the serialized size on the first pass. Then later in the callback, SerializeResponseOutput will query the same metadata from the response object again, but it's not particularly expensive as it doesn't copy the output buffer. We could store the response metadata in some struct on the first pass and re-use that on the second pass, but I'm not sure if that's any more efficient that just re-using the InferenceResponse output as the "struct" that's holding the metadata on the second pass.

Let me know if that makes sense or not

File: src/cache_manager.cc [hunk 2]
```diff
@@ -323,25 +369,11 @@ Status
...
-            ", received: " + std::to_string(items.size()));
-  }
-
-  for (size_t i = 0; i < items.size(); i++) {
-    items[i]->ToResponse(responses[i]);
-  }
-
+  auto lentry = std::make_unique<CacheEntry>();
+  // Create response allocator to copy directly from cache to response buffers
+  auto allocator = CacheToResponseAllocator(responses);
```
[the remark this answers is not in the document]
Comment (rmccorm4) [reply src/cache_manager.cc:431]: This shouldn't be making a copy of the object, only a copy of the pointer. I added a CopyConstructor for every allocator that prints out when its called to verify this, and it was never called on insert or lookup.

File: src/cache_entry.cc [hunk 3]
```diff
@@ -32,123 +32,305 @@ namespace triton { namespace core {
...
+
+  for (size_t i = 0; i < responses.size(); i++) {
+    RETURN_IF_ERROR(DeserializeBuffer(responses[i], buffers_[i]));
+  }
+
+  return Status::Success;
+}
+
+Status
+CacheEntry::DeserializeBuffer(InferenceResponse* response, Buffer buffer)
```
[the remark this answers is not in the document]
Comment (rmccorm4) [reply src/cache_entry.cc:308]: cleaned this up

File: src/cache_entry.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (Tabrizian) [reply src/cache_entry.cc:323]: Ah, I see. I mistaken the metadata copies with actual buffer copy. I think that should be fine. Resolving.

**Review by Tabrizian — APPROVED** [review]

**Pull request merged**